### PR TITLE
aws: fix for webidentity expiration mismatch with sts response

### PR DESCRIPTION
**Pull request by nbaws**
Commit Message: aws: fix for webidentity expiration mismatch with sts response
Additional Description: sts response contains scientific notation, this patch handles that expiration case
Risk Level: Low
Testing: Unit
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] https://github.com/envoyproxy/envoy/issues/33806
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/extensions/common/aws/utility_test.cc [hunk 1]
```diff
@@ -495,26 +495,57 @@ TEST(UtilityTest, JsonStringFound) {
...
+  auto test_json = Json::Factory::loadFromStringNoThrow("{\"expiration\":-0.17144492389}");
+  EXPECT_TRUE(test_json.ok());
+  auto value_or_error = test_json.value()->getValue("expiration");
+  EXPECT_TRUE(value_or_error.ok());
+  EXPECT_FALSE(absl::holds_alternative<int64_t>(value_or_error.value()));
+  EXPECT_TRUE(absl::holds_alternative<double>(value_or_error.value()));
+  const auto expiration =
+      Utility::getIntegerFromJsonOrDefault(test_json.value(), "expiration", 9999);
+  // Should return default value
+  EXPECT_EQ(expiration, 9999);
```
Comment (suniltheta) [inline test/extensions/common/aws/utility_test.cc:546]: how this specific value?

File: test/extensions/common/aws/credentials_provider_impl_test.cc [hunk 2]
```diff
@@ -2311,6 +2311,38 @@ TEST_F(WebIdentityCredentialsProviderTest, FullCachedCredentialsWithMissingExpir
...
+  const auto credentials = provider_->getCredentials();
+  EXPECT_EQ("akid", credentials.accessKeyId().value());
+  EXPECT_EQ("secret", credentials.secretAccessKey().value());
+  EXPECT_EQ("token", credentials.sessionToken().value());
+}
+
+TEST_F(WebIdentityCredentialsProviderTest, RefreshOnNormalCredentialExpirationIntegerFormat) {
   // Setup timer.
   timer_ = new NiceMock<Event::MockTimer>(&context_.dispatcher_);
   // Time 2018-01-02T05:04:05Z in unix_timestamp is 1514869445
```
Comment (suniltheta) [inline test/extensions/common/aws/credentials_provider_impl_test.cc:2348]: Is it possible to get integer expiration format line in L2357 ?
Are we making the functionality to handle both?

Comment (suniltheta) [reply test/extensions/common/aws/credentials_provider_impl_test.cc:2348]: If we are sure STS json response will always contain "scientific notation" then we can get rid of `RefreshOnNormalCredentialExpirationIntegerFormat` test case, assume that as bad format and ignore the expiration value and depend on default refresh period of 1 hour.

File: test/extensions/common/aws/utility_test.cc (hunk 1 above)
Comment (nbaws) [reply test/extensions/common/aws/utility_test.cc:546]: test case puts an invalid negative number in and expects the default value to be returned. default value is set to 9999 for this test

File: test/extensions/common/aws/credentials_provider_impl_test.cc (hunk 2 above)
Comment (nbaws) [reply test/extensions/common/aws/credentials_provider_impl_test.cc:2348]: utility.cc will handle both cases. I don't expect that STS will put a fix in for normal integers any time soon, but if they do we will continue to function correctly.

i put the RefreshOnNormalCredentialExpirationIntegerFormat test case to ensure both scenarios will always work as expected

**Review by suniltheta — APPROVED** [review]

**Comment by nbaws** [comment]
thanks for the review @suniltheta

**Review by jmarantz — APPROVED** [review]

**Pull request merged**